docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
+    
+### Security Context Provider
+
+The Kubelet will have an interface that points to a `SecurityContextProvider`. The `SecurityContextProvider` is invoked before creating and running a given container:
+
+```go
+type SecurityContextProvider interface {
+    // ModifyContainerConfig is called before the Docker createContainer call.
+    // The security context provider can make changes to the Config with which
+    // the container is created.
+    // An error is returned if it's not possible to secure the container as 
+    // requested with a security context. 
+	ModifyContainerConfig(pod *api.BoundPod, container *api.Container, config *docker.Config) error
+	
+	// ModifyHostConfig is called before the Docker runContainer call.
+	// The security context provider can make changes to the HostConfig, affecting
+	// security options, whether the container is privileged, volume binds, etc.
+	// An error is returned if it's not possible to secure the container as requested 
+    // with a security context. 
+	ModifyHostConfig(pod *api.BoundPod, container *api.Container, hostConfig *docker.HostConfig)
+}
+```
+If the value of the SecurityContextProvider field on the Kubelet is nil, the kubelet will create and run the container as it does today.   
+
+### Security Context
+
+A security context has a 1:1 correspondence to a service account and it can be included as
+part of the service account resource. Following is an example of an initial implementation:
+
+```go
+type SecurityContext struct {
+    // user is the uid to use when running the container
+	User int
+	
+	// allowPrivileged indicates whether this context allows privileged mode containers
+	AllowPrivileged bool
+	
+	// allowedVolumeTypes lists the types of volumes that a container can bind
+	AllowedVolumeTypes []string
+	
+	// addCapabilities is the list of Linux kernel capabilities to add
+	AddCapabilities []string
+	
+	// removeCapabilities is the list of Linux kernel capabilities to remove
+	RemoveCapabilities []string
+	
+	// SELinux specific settings (optional)
+	SELinux *SELinuxContext
+	
+	// AppArmor specific settings (optional)
+	AppArmor *AppArmorContext
+	
+	// FUTURE:
+	// With Linux user namespace support, it should be possible to map
+	// a range of container uids/gids to arbitrary host uids/gids
+	// UserMappings []IDMapping
+	// GroupMappings []IDMapping
+}
+
+type SELinuxContext struct {
+    // MCS label/SELinux level to run the container under
+    Level string
+    
+    // SELinux type label for container processes
+    Type  string    
+    
+    // FUTURE:
+    // LabelVolumeMountsExclusive []Volume
+    // LabelVolumeMountsShared    []Volume
+}
+
+type AppArmorContext struct {
+	// AppArmor profile
+	Profile string
+}
```

Comment:
----- Original Message -----
> > +type SELinuxContext struct {
> > +    // MCS label/SELinux level to run the container under
> > +    Level string
> > +
> > +    // SELinux type label for container processes
> > +    Type  string
> > +
> > +    // FUTURE:
> > +    // LabelVolumeMountsExclusive []Volume
> > +    // LabelVolumeMountsShared    []Volume
> > +}
> > +
> > +type AppArmorContext struct {
> > +	// AppArmor profile
> > +	Profile string
> > +}
> 
> I see developers as:
>   - writing pod specs.
>   - making some of their own images, and using some "community" docker images
>   - knowing which pods need to talk to which other pods
>   - decides which pods should be share files with other pods, and which
>   should not.
>   - reasons about application level security, such as containing the effects
>   of a local-file-read exploit in a webserver pod.
>   - does not often reason about operating system or organizational security.
>   - not necessarily comfortable reasoning about the security properties of a
>   system at the level of detail of Linux Capabilities, SELinux, AppArmor,
>   etc.
>   
> I see project admins as:
>   - allocating identities and roles and namespaces.
>   - reasoning about organizational security:
>     - don't give a developer permissions that are not needed for role.

Also: Concerned about how some things running as higher trust (builds can push images to docker repo X) don't get abused by ordinary developers.  Your phrasing is fine, just adding a scenario we think about.


>     - protect files on shared storage from unnecessary cross-team access
>   - less focused about application security

> 
> I see cluster admins as:
>   - less focused on application security.  Focused on operation system
>   security.
>   - protects the node from bad actors in containers, and properly-configured
>   innocent containers from bad actors in other containers.
>   -  comfortable reasoning about the security properties of a system at the
>   level of detail of Linux Capabilities, SELinux, AppArmor, etc.
>   - decides who can use which Linux Capabilities, run privileged containers,
>   use hostDir, etc.
>     - e.g. a team that manages Ceph or a mysql server might be trusted to
>     have raw access to storage devices in some organizations, but teams that
>     develop the applications at higher layers would not.
> 
> Do you agree that those are reasonable separations of responsibilities for a
> Kubernetes cluster?

Yes, nailed it.  Those distinctions should be in this proposal and in service account (or whatever context it takes)

> 
> If so, do you think the current design allows those three groups to work
> independently of each other and to focus on the information they need?  I'm
> not sure; I'm trying to think that through.

I think the proposal doesn't describe the higher level pieces that are in service account and secrets, but assumes they exist.  I would feel like service account is a concept for the developer end of the spectrum and security context is much more about the other end.  The cluster and project admins must allow developers to have capabilities, the developers must understand how they use those capabilities, and in general higher level developer concepts get boiled down into security contexts and execution details.  So this proposal is definitely talking about a part of the overall story.

The outcome of these proposals / prototypes should at minimum include a document that describes the above and how the pieces provide that spectrum.

I *think* at this point that I could argue a convincing story about:

* how namespaces get a service account by default (configured by cluster admins)
* project admins can tweak both those and authorization policies (what we are working through here https://github.com/openshift/origin/blob/master/docs/proposals/policy.md) to properly isolate those
* some secrets get added by default or via developers directly adding them to service accounts or their pods
* how service accounts are converted to a security context down to the kubelet via finalizers / admission controller
* how secrets from a service account (or other mechanism) could be bind mounted into containers, either a la the docker vault / secrets proposals, or via a more specific volume type
* how the kubelet could take the info on the security context and turn it into a user namespace / labels / volumes
* how developers could adapt their images and applications to work within the limitations of the above items

There is of course a lot of handwaving in between those bits.